embedded-graphics/src/primitives/circle.rs

```diff
@@ -177,91 +163,84 @@ impl Points {
-        loop {
-            let len = (self.c.x - 2 * self.p.x).pow(2) + (self.c.y - 2 * self.p.y).pow(2);
+        let threshold = self.threshold;
+        self.iter
+            .find(|(_, distance)| *distance < threshold)
+            .map(|(point, _)| point)
+    }
+}
 
-            let point = if len < self.threshold {
-                Some(self.p)
-            } else {
-                None
-            };
+/// Pixel iterator for each pixel in the circle border
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct StyledCircleIterator<C>
+where
+    C: PixelColor,
+{
+    iter: DistanceIterator,
 
-            self.p.x += 1;
+    outer_threshold: u32,
+    outer_color: Option<C>,
 
-            if self.p.x > self.top_left.x + self.diameter as i32 {
-                self.p.x = self.top_left.x;
-                self.p.y += 1;
-            }
+    inner_threshold: u32,
+    inner_color: Option<C>,
+}
 
-            if self.p.y > self.top_left.y + self.diameter as i32 {
-                break None;
-            }
+impl<C> StyledCircleIterator<C>
+where
+    C: PixelColor,
+{
+    fn new(styled: &Styled<Circle, PrimitiveStyle<C>>) -> Self {
+        // Always use a stroke width of 0 if no stroke color was set.
+        let stroke_width = match styled.style.stroke_color {
+            Some(_) => styled.style.stroke_width,
+            None => 0,
+        };
 
-            if point.is_some() {
-                break point;
-            }
+        let outer_diameter = styled.primitive.diameter;
+        let inner_diameter = outer_diameter.saturating_sub(2 * stroke_width);
+
+        let inner_threshold = diameter_to_threshold(inner_diameter);
+        let outer_threshold = diameter_to_threshold(outer_diameter);
+
+        Self {
+            iter: DistanceIterator::new(&styled.primitive),
+            outer_threshold,
+            outer_color: styled.style.stroke_color,
+            inner_threshold,
+            inner_color: styled.style.fill_color,
         }
     }
 }
 
-/// Pixel iterator for each pixel in the circle border
-#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
-pub struct StyledCircleIterator<C: PixelColor> {
-    top_left: Point,
-    diameter: u32,
-    style: PrimitiveStyle<C>,
-    p: Point,
-    c: Point,
-    outer_threshold: i32,
-    inner_threshold: i32,
-}
-
 impl<C> Iterator for StyledCircleIterator<C>
 where
     C: PixelColor,
 {
     type Item = Pixel<C>;
 
-    // https://stackoverflow.com/a/1237519/383609
-    // https://stackoverflow.com/questions/1201200/fast-algorithm-for-drawing-filled-circles#comment80182898_1237519
     fn next(&mut self) -> Option<Self::Item> {
         // If the fill and stroke colors are `None`, treat entire object as transparent and exit
         // early.
-        if self.style.stroke_color.is_none() && self.style.fill_color.is_none() {
+        if self.outer_color.is_none() && self.inner_color.is_none() {
             return None;
         }
```

Comment:
This might be a bit more pleasant:

```suggestion
        self.outer_color.or(self.inner_color)?;
```

This also optimises to like 2 fewer asm instructions so yay I guess :grinning:

Reply:
That's much nicer. But I've decided to remove it completely, which should remove some additional asm instructions, but i haven't counted :wink: